src/core/MIGRATION.md

```diff
@@ -0,0 +1,683 @@
+  * [Move all webpack alias imports into uiExport entry files](#move-all-webpack-alias-imports-into-uiexport-entry-files)
+  * [Switch to new platform services](#switch-to-new-platform-services-1)
+  * [Migrate to the new plugin system](#migrate-to-the-new-plugin-system-1)
+* [Frequently asked questions](#frequently-asked-questions)
+  * [Is migrating a plugin an all-or-nothing thing?](#is-migrating-a-plugin-an-all-or-nothing-thing)
+  * [Do plugins need to be converted to TypeScript?](#do-plugins-need-to-be-converted-to-typescript)
+  * [How is static code shared between plugins?](#how-is-static-code-shared-between-plugins)
+  * [How is "common" code shared on both the client and server?](#how-is-common-code-shared-on-both-the-client-and-server)
+  * [When does code go into a plugin, core, or packages?](#when-does-code-go-into-a-plugin-core-or-packages)
+
+Make no mistake, it is going to take a lot of work to move certain plugins to the new platform. Our target is to migrate the entire repo over to the new platform throughout 7.x and to remove the legacy plugin system no later than 8.0, and this is only possible if teams start on the effort now.
+
+The goal of this document is to guide teams through the recommended process of migrating at a high level. Every plugin is different, so teams should tweak this plan based on their unique requirements.
+
+We'll start with an overview of how plugins work in the new platform, and we'll end with a generic plan of action that can be applied to any plugin in the repo today.
+
+## Overview
+
+Plugins in the new platform are not especially novel or complicated to describe. Our intention wasn't to build some clever system that magically solved problems through abstractions and layers of obscurity, and we wanted to make sure plugins could continue to use most of the same technologies they use today, at least from a technical perspective.
+
+New platform plugins exist in the `src/plugins` and `x-pack/plugins` directories.
+
+### Architecture
+
+Plugins are defined as classes and exposed to the platform itself through a simple wrapper function. A plugin can have browser side code, server side code, or both. There is no architectural difference between a plugin in the browser and a plugin on the server, which is to say that in both places you describe your plugin similarly, and you interact with core and/or other plugins in the same way.
+
+The basic file structure of a new platform plugin named "demo" that had both client-side and server-side code would be:
+
+```
+src/plugins
+  demo
+    kibana.json [1]
+    public
+      index.ts [2]
+      plugin.ts [3]
+    server
+      index.ts [4]
+      plugin.ts [5]
+```
+
+**[1] `kibana.json`** is a static manifest file that is used to identify the plugin and to determine what kind of code the platform should execute from the plugin:
+
+```json
+{
+  "id": "demo",
+  "server": true,
+  "ui": true
```

Comment:
I think it's much easier to understand and build up a mental model if the configuration option key is consistent with other naming conventions. Both should be `ui` or both `public` etc.

To dig up an old discussion, although `/public` is a strong convention for a folder structure, it's much less intuitive as a configuration option or as code e.g. `PublicStartContract` or `"public": true`(appears to be a visibility setting?). It's also unnatural to talk about "public code", we're more likely to say "client-side code" or "ui code".

So I'd prefer if we change `public` -> `ui`.

Reply:
I'm going to leave this as-is in the guide for now since `ui` is what exists, but feel free to open an issue to discuss this separately.